### PR TITLE
Host cleanup

### PR DESCRIPTION
**Pull request by lukechampine**
Added a logger and various minor improvements. Notably, the host now saves the full revision transaction, instead of just modifying the original file contract object.
I'm 90% sure this won't cause compatibility problems, but I'm running this code on my local hosting node for a while to make sure. (Host could use a lot more tests...)

### CODE REVIEW

File: modules/consensus/consensusset.go [hunk 1]
```diff
@@ -142,7 +142,7 @@ func (cs *ConsensusSet) BlockAtHeight(height types.BlockHeight) (block types.Blo
 		exists = true
 		return nil
 	})
-	return block, exists
+	return
```
Comment (DavidVorick) [inline modules/consensus/consensusset.go]: not fan

File: modules/consensus/consensusset.go [hunk 2]
```diff
@@ -204,7 +204,7 @@ func (cs *ConsensusSet) InCurrentPath(id types.BlockID) (inPath bool) {
 		inPath = pathID == id
 		return nil
 	})
-	return inPath
+	return
```
Comment (DavidVorick) [inline modules/consensus/consensusset.go]: still not fan

File: modules/gateway/gateway_test.go [hunk 3]
```diff
@@ -32,7 +32,7 @@ func TestAddress(t *testing.T) {
 		t.Fatal("Address does not return g.myAddr")
 	}
 	port := modules.NetAddress(g.listener.Addr().String()).Port()
-	expAddr := modules.NetAddress(net.JoinHostPort("::1", port))
+	expAddr := modules.NetAddress(net.JoinHostPort("::", port))
```
Comment (DavidVorick) [inline modules/gateway/gateway_test.go]: ?

File: modules/host/announce.go [hunk 4]
```diff
@@ -70,16 +56,15 @@ func (h *Host) Announce() error {
 	addr := h.myAddr
 	h.mu.RUnlock(lockID)
 
-	// Check that the host's ip address is both known and reachable.
-	if addr.Host() == "::1" {
+	// Check that the host's ip address is known.
+	if addr.IsLocal() {
```
Comment (DavidVorick) [inline modules/host/announce.go]: does this also catch things like 192.168?

File: modules/host/host.go [hunk 5]
```diff
@@ -2,13 +2,12 @@ package host
...
 import (
 	"errors"
+	"log"
 	"net"
-	"os"
 	"sync"
 
 	"github.com/NebulousLabs/Sia/crypto"
 	"github.com/NebulousLabs/Sia/modules"
-	"github.com/NebulousLabs/Sia/modules/consensus"
```
Comment (DavidVorick) [inline modules/host/host.go]: woo

File: modules/host/host.go [hunk 6]
```diff
@@ -2,13 +2,12 @@ package host
...
 	safesync "github.com/NebulousLabs/Sia/sync"
```
Comment (DavidVorick) [inline modules/host/host.go]: probably okay to move the whole package to just "sync", safesync probably not needed anymore

File: modules/host/negotiate.go [hunk 7]
```diff
@@ -16,6 +16,11 @@ import (
 	"github.com/NebulousLabs/Sia/types"
 )
 
+const (
+	// maxRevisionSize is the maximum number of bytes added in a single revision
+	maxRevisionSize = 100e6 // 100 MB
```
Comment (DavidVorick) [inline modules/host/negotiate.go]: 100e6 -> one of the first times I actually had trouble figuring out what something said on my small test. Didn't help that the '0's had dots in the middle

File: modules/host/negotiate.go [hunk 8]
```diff
@@ -230,7 +239,7 @@ func (h *Host) rpcUpload(conn net.Conn) error {
 
 	// send doubly-signed transaction set
 	if err := encoding.WriteObject(conn, signedTxnSet); err != nil {
-		return err
+		return errors.New("couldn't write signed transaction set: " + err.Error())
```
Comment (DavidVorick) [inline modules/host/negotiate.go]: ah, wonderful

File: modules/netaddress.go [hunk 9]
```diff
@@ -39,7 +39,7 @@ func (na NetAddress) IsLocal() bool {
 	if ip := net.ParseIP(host); ip != nil && ip.IsLoopback() {
 		return true
 	}
-	if host == "localhost" || host == "0.0.0.0" {
+	if host == "localhost" || host == "::" || host == "0.0.0.0" {
```
Comment (DavidVorick) [inline modules/netaddress.go]: the ever growing blacklist :P

**Comment by DavidVorick** [comment]
I'm feeling increasingly strongly about needing an integration test that will make sure the host is getting paid and is submitting the file contract correctly. Too much changed here and I don't feel like our test suite is strong enough for me to be confident that nothing broke.

Hosting is one of the most important parts of the network, it makes sense to throw a ton of testing at it.

File: modules/gateway/gateway_test.go (hunk 3 above)
Comment (lukechampine) [inline modules/gateway/gateway_test.go]: consequence of a5007062cd

File: modules/host/announce.go (hunk 4 above)
Comment (lukechampine) [inline modules/host/announce.go]: no, only `127.0.0.1` and `0.0.0.0`

Comment (DavidVorick) [inline modules/host/announce.go]: yeah probably, or extend the function to include the local ips

**Pull request merged**